drivers/flash/spi_flash_sfdp.c

```diff
@@ -1,558 +1,993 @@
+		}
 
+		SYS_LOG_DBG("erase instruction: %02x address: %08x",
+			    erase_opcode, offs);
+		ret = spi_flash_erase_cmd(dev, erase_opcode, offs);
+		offs += erase_size;
+		left -= erase_size;
+	}
+	SYS_LOG_INF("erase: %08x-%08x", offset, offset + size - 1);
+end_erase:
 	k_sem_give(&data->sem);
 	return ret;
 }
 
+static int spi_flash_write_protection_set(struct device *dev, bool enable)
+{
+	struct spi_flash_data *const data = dev->driver_data;
+
+	if (data->write_protection) {
+		return data->write_protection(dev, enable);
+	}
+	data->write_protection_sw = enable;
+	return 0;
+}
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static void spi_flash_pages_layout(struct device *dev,
+				   const struct flash_pages_layout **layout,
+				   size_t *layout_size)
+{
+	struct spi_flash_data *const data = dev->driver_data;
+
+	if (layout) {
+		*layout = data->pages_layouts;
+	}
+	if (layout_size) {
+		*layout_size = data->pages_layouts_count;
+	}
+}
+#endif
+
 static const struct flash_driver_api spi_flash_api = {
 	.read = spi_flash_read,
 	.write = spi_flash_write,
 	.erase = spi_flash_erase,
 	.write_protection = spi_flash_write_protection_set,
 #if defined(CONFIG_FLASH_PAGE_LAYOUT)
-	.page_layout = (flash_api_pages_layout)
-		       flash_page_layout_not_implemented,
+	.page_layout = spi_flash_pages_layout,
 #endif
 	.write_block_size = 1,
 };
 
+static int spi_flash_search_sector_map(struct device *dev, dword_t *smpt,
+				       int len)
+{
+	u8_t id = 0;
+	struct spi_flash_data *data = dev->driver_data;
+	union sfdp_sector_map_parameters *smp = (void *)smpt;
+
+	while (len > 0) {
+		if (smp->type) {
+			if (smp->map.id == id) {
+				if (smp->map.region_count + 1 >
+				    ARRAY_SIZE(data->smrp)) {
+					SYS_LOG_ERR("SMRP array size is too "
+						    "small");
+					return -ENODEV;
+				}
+				data->smrp_count = smp->map.region_count + 1;
+				memcpy(data->smrp, smp + 1,
+				       sizeof(data->smrp[0]) *
+					   data->smrp_count);
+				SYS_LOG_INF("Sector Map finish !");
+				return 0;
+			}
+			len -= smp->map.region_count + 2;
+			smp += smp->map.region_count + 2;
+		} else {
+			/* Configuration Detection */
+			/* todo */
+			SYS_LOG_WRN(
+			    "Go to do sector map Configuration Detection");
+			len -= 2;
+			smp += 2;
+		}
+		if (smp->sequence_end) {
+			SYS_LOG_INF("Sector Map finish !");
+			return 0;
+		}
+	}
+	SYS_LOG_ERR("Wrong Sector Map Parameters");
+	return -ENODEV;
+}
+
 static int spi_flash_init(struct device *dev)
 {
-	struct device *spi_dev;
 	struct spi_flash_data *data = dev->driver_data;
+	struct spi_flash_init_config init_config = {0};
 	int r;
 
-	spi_dev = device_get_binding(CONFIG_SPI_FLASH_SFDP_SPI_NAME);
-	if (!spi_dev) {
+	data->spi = device_get_binding(CONFIG_SPI_FLASH_SPI_NAME);
+	if (!data->spi) {
 		return -EIO;
 	}
 
-	data->config = (struct spi_config) {
-		.dev = spi_dev,
-		.frequency = CONFIG_SPI_FLASH_SFDP_SPI_FREQ_0,
-		.operation = SPI_WORD_SET(8),
-	};
+#if defined(SPI_FLASH_GPIO_SPI_CS)
```

Comment:
Why not a Kconfig option for this? So it would be CONFIG_ prefixed then.

Reply:
oh...
This is a lapse